### DISCUSSION TITLE
v0.4.0 Testing Feedback

### DISCUSSION THREAD

**Discussion by lee1043** [xCDAT/xcdat#391] [4. General]
Discussion related to v0.4.0

**Comment by lee1043** [comment]
Was `decode_non_cf_time` deprecated? I got `AttributeError: module 'xcdat' has no attribute 'decode_non_cf_time'` message while running my code after updating to v0.4.0.

**Reply by pochedls** [reply]
It was deprecated [here](https://github.com/xCDAT/xcdat/commit/c6587b88c3d6d834d4a821890ad2af80e77d000d) because we handle decoding of CF and non-CF datasets with one function now.

**Reply by lee1043** [reply]
@pochedls good to know, thank you for clarifying.

**Reply by tomvothecoder** [reply]
Renaming `decode_non_cf_time()` to `decode_time()` is technically a breaking change that warranted a major release. 
However, I assumed direct usage of this function was low and we haven't released a v1.0.0 yet.

Based on this discussion, we should consider expediting the v1.0.0 release soon (#375) if there are any more breaking changes and/or when #388 is merged.

**Comment by lee1043** [comment]
Following code was working okay in v0.3.3 but reporting the following error in v0.4.0, which I am curious if the following behavior is expected / desired or if it is a bug. 

```python
mod_file = "cmip6.historical.E3SMv2.r1i1p1.mon.pr.198101-200512.AC.v20221027_test2-regridded.nc"  # xCDAT calculated AC and regridded
dm_x = xc.open_dataset(mod_file, data_var="pr")
dm_x.temporal.average(var)
```

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File /opt/anaconda3/envs/pcmdi_metrics_dev_5/lib/python3.9/site-packages/xcdat/temporal.py:733, in TemporalAccessor._set_data_var_attrs(self, data_var)
    732 try:
--> 733     self.calendar = dv[self.dim].encoding["calendar"]
    734     self.date_type = get_date_type(self.calendar)

KeyError: 'calendar'

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
Cell In [26], line 1
----> 1 dm_x.temporal.average(var)

File /opt/anaconda3/envs/pcmdi_metrics_dev_5/lib/python3.9/site-packages/xcdat/temporal.py:194, in TemporalAccessor.average(self, data_var, weighted, keep_weights)
    149 def average(self, data_var: str, weighted: bool = True, keep_weights: bool = False):
    150     """
    151     Returns a Dataset with the average of a data variable and the time
    152     dimension removed.
   (...)
    192     >>> ds_month.ts
    193     """
--> 194     self._set_data_var_attrs(data_var)
    196     freq = self._infer_freq()
    198     return self._averager(data_var, "average", freq, weighted, keep_weights)

File /opt/anaconda3/envs/pcmdi_metrics_dev_5/lib/python3.9/site-packages/xcdat/temporal.py:736, in TemporalAccessor._set_data_var_attrs(self, data_var)
    734     self.date_type = get_date_type(self.calendar)
    735 except KeyError:
--> 736     raise KeyError(
    737         f"The 'calendar' encoding attribute is not set on the '{data_var}' "
    738         f"time coordinate variable ({self.dim}). This might indicate that the "
    739         "time coordinates were not decoded, which is required for temporal "
    740         "averaging operations. "
    741     )

KeyError: "The 'calendar' encoding attribute is not set on the 'pr' time coordinate variable (time). This might indicate that the time coordinates were not decoded, which is required for temporal averaging operations. "
```

**Reply by pochedls** [reply]
It seems like in v0.4 when setting the time attributes an explicit [error is thrown](https://github.com/xCDAT/xcdat/blame/c6587b88c3d6d834d4a821890ad2af80e77d000d/xcdat/temporal.py#L711) if the calendar is missing. 

I would have thought that xcdat would put `standard` in for calendar on open, but maybe it simply guesses the calendar when decoding the time. I think either way, we don't want it to fail here if the time axis is already decoded (it doesn't need the calendar to do temporal averaging).

**Reply by tomvothecoder** [reply]
`decode_time()` defaults to `calendar="standard"` if the `calendar` attr is missing.
https://github.com/xCDAT/xcdat/blob/d5d4fd77cbe0e23cc6452c2a1d421923d800b31b/xcdat/dataset.py#L318-L324

In your Dataset, I'm curious as to why the `calendar` attribute is missing after opening and decoding.

Also, I should replace this check to see if time coordinates are decoded:
https://github.com/xCDAT/xcdat/blob/d5d4fd77cbe0e23cc6452c2a1d421923d800b31b/xcdat/temporal.py#L732-L741

with this function:
https://github.com/xCDAT/xcdat/blob/d5d4fd77cbe0e23cc6452c2a1d421923d800b31b/xcdat/dataset.py#L491-L509

Can you provide this file for us to debug?

**Reply by lee1043** [reply]
@tomvothecoder @pochedls Thank you for your comment. It looks like I have made a mistake when handling the data. In the original code I was opening two files: one that cdms written another written using xcdat. And I was manually copying the time dimension from cdms written dataset to xcdat written dataset, then applied temporal average to the xcdat written dataset. My intention was to correct time mismatching in annual cycle. The issue was resolved when I copied time dimension in the opposite way; copy time from xcdat-written dataset to the cdms-written dataset. My apology for the confusion.

**Reply by tomvothecoder** [reply]
@lee1043 No worries, that clears things up

**Comment by lee1043** [comment]
Just confirming that v0.4.0 update resulted no change in the PMP mean climate statistics (bias, rmse, correlation, standard deviation, etc), which is GREAT because we don't want these numbers changing with the newer version!

**Reply by tomvothecoder** [reply]
Awesome, I'm glad to hear things are identical between versions! Thanks for testing Jiwoo!

**Reply by tomvothecoder** [reply]
Also thanks for opening up this new thread since #296 is outdated.

**Comment by arfriedman** [comment]
Hi all,

I am a recent xcdat user. I've encountered a specific problem temporal averaging a dataset with a fixed-frequency cftime axis which seems related to the calendar issue discussed here.

I am examining monthly data from ERSST:
```
import cftime
import xarray as xr
import xcdat as xc
ERSST = xc.open_dataset("http://psl.noaa.gov/thredds/dodsC/Datasets/noaa.ersst.v5/sst.mnmean.nc")
```
I change the time to a fixed-frequency cftime range (to allow for comparison with other datasets in my project):
`ERSST["time"] = xr.cftime_range(start=cftime.datetime(1854, 1, 1), end=cftime.datetime(2022, 12, 31), freq="M")`

Sseasonal averaging works fine with the original time axis (not shown), but attempting to take a seasonal average of the modified fixed-cftime-frequency dataset results in an error:
```
ERSST_seas = ERSST.temporal.group_average("sst", "season")
KeyError: "The 'calendar' encoding attribute is not set on the 'sst' time coordinate variable (time). This might indicate that the time coordinates were not decoded, which is required for temporal averaging operations. "
```

I just wanted to check if the error with the modified axis is a bug, or if I am pushing the software to the limit.

Thanks much,
Andrew

**Reply by pochedls** [reply]
This is similar to [this discussion](https://github.com/xCDAT/xcdat/discussions/381#discussioncomment-4041530). I think you can add the calendar attribute:

```
import xcdat as xc
import xarray as xr
import cftime

ds = xc.open_dataset("http://psl.noaa.gov/thredds/dodsC/Datasets/noaa.ersst.v5/sst.mnmean.nc")
ds["time"] = xr.cftime_range(start=cftime.datetime(1854, 1, 1), end=cftime.datetime(2022, 12, 31), freq="M")
encoding = ds.time.encoding
encoding['calendar'] = 'standard'
```

But when the custom time axis is added, it doesn't have time bounds (which are needed for group averaging): 

> KeyError: "No bounds were found for the 'T' axis. Make sure bounds vars exist in the Dataset with names that match the 'bounds' keys, or try adding bounds."

You can fix this with: 

```
ds = ds.bounds.add_missing_bounds()
```

At some point, we should investigate whether the first `KeyError` is needed. I think this was probably used to figure out if the time axis could be used for this computation, but the `calendar` attribute probably isn't actually required. So the logic could possibly be updated here.

**Reply by tomvothecoder** [reply]
Generating time coordinates using pandas/xarray and using them with xCDAT temporal APIs seem to be a desired functionality.
We should look more closely at how we can better support this workflow (e.g., better documentation, alternative solutions, etc.). I'll open up an issue and link it [here](https://github.com/xCDAT/xcdat/issues/414).

> At some point, we should investigate whether the first KeyError is needed. I think this was probably used to figure out if the time axis could be used for this computation, but the calendar attribute probably isn't actually required. So the logic could possibly be updated here.

When decoding time, [xarray.decode_cf_datetime](https://github.com/pydata/xarray/blob/b4515582ffc8b7f63632bfccd109d19889d00384/xarray/coding/times.py#L281-L283) defaults `calendar = "standard"` if the `calendar` encoding attribute is not set.

We also adopted this behavior in `xcdat.decode_time`: https://github.com/xCDAT/xcdat/blob/16a807cb8c885700efc85d3223b8392f3627bbc3/xcdat/dataset.py#L318-L324

We can consider doing the same thing in the temporal APIs and raise a warning for a missing calendar attribute (instead of `KeyError`).

**Reply by pochedls** [reply]
@tomvothecoder – but do we actually *need* the calendar attribute? I'm guessing this isn't actually needed for group_averaging (and so we don't need a warning / error for this specific issue). Maybe we don't need the warning or the error could be more targeted?

**Reply by tomvothecoder** [reply]
@pochedls we have preprocessing logic to drop leap days when calculating climatology/departures. This logic depends on the calendar type:
https://github.com/xCDAT/xcdat/blob/16a807cb8c885700efc85d3223b8392f3627bbc3/xcdat/temporal.py#L889-L894

We don't have any calendar specific logic for `.average` and `.group_average`. I think the error could be more targeted instead of generalized across all of the APIs.

**Reply by tomvothecoder** [reply]
I'll include this information in #414.

**Reply by arfriedman** [reply]
Thank you @tomvothecoder and @pochedls

**Reply by pochedls** [reply]
@arfriedman – I realize that the workaround that I gave you is problematic because of [this issue](https://github.com/xCDAT/xcdat/issues/412). Basically – since the original time bounds were dropped (and then generated with `.add_missing_bounds()`) they will not correspond to the calendar correctly. I think the effect of this issue is that your seasonal averages will not receive correct weights (so instead of February receiving 28 days of "weight" it probably is receiving 30 days).

**Reply by arfriedman** [reply]
Thank you @pochedls for letting me know. I see that #412 would definitely be helpful.

**Reply by tomvothecoder** [reply]
#414 improves the usage of custom time coordinates with the temporal accessor APIs.

Instead of raising an error, the temporal accessor APIs will raise a warning if a "calendar" attribute is not set and defaults it to "standard". The dataset still needs time bounds before the APIs are called.

**Comment by jypeter** [comment]
That's a good example where some equivalent of `setTimeBoundsMonthly` would be useful :-)
